### PR TITLE
Cloud Carbon Footprint for On-Premise post

### PR DESCRIPTION
**Pull request by mgriffin-scottlogic**
Please add a direct link to your post here:

https://mgriffin-scottlogic.github.io/blog/2024/06/20/cloud-carbon-footprint-on-premise.html

Have you (please tick each box to show completion):

 - [x] Added your blog post to a single category?
 - [x] Added a brief summary for your post? Summaries should be roughly two sentences in length and give potential readers a good idea of the contents of your post.
 - [x] Checked that the build passes?
 - [x] Checked your spelling (you can use `npm install` followed by `npx mdspell "**/{FILE_NAME}.md" --en-gb -a -n -x -t` if that's your thing)
 - [x] Ensured that your author profile contains a profile image, and a brief description of yourself? (make it more interesting than just your job title!)
 - [x] Optimised any images in your post? They should be less than 100KBytes as a general guide.

Posts are reviewed / approved by your Regional Tech Lead.

### CODE REVIEW

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+---
+
+## Introduction
+
+The Cloud Carbon Footprint tool was discussed in a [previous blog post](https://blog.scottlogic.com/2023/10/19/tools-for-measuring-cloud-carbon-emissions.html), as a third party method of estimating emissions associated with cloud workloads. However, it can also use data from [On-Premise](https://www.cloudcarbonfootprint.org/docs/on-premise) sources for its estimations. This blog post gives an overview of how it should be used and investigates how it works via testing and analysis of its code. I consider the issues with its use and list some contributions that could be made to improve it.
+
+## Usage Overview
+
+The system requires a .csv file as an input, which must contain a minimum set of columns, with Machine Name, cost, power usage effectiveness and CPU utilization being optional. If CPU Utilization is not specified then it is assumed to be 50% by default, which can also be configured to use a different amount for server, desktop, and laptop machine types.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Is "utilization" capitalised, as in the second sentence, or not, as in the first?
At "default", either: "default; it can also..." or "default, which can be...".

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+This .csv file must then be passed to a command, which generates an output .csv file with estimated energy usage and CO2e emissions. This command is `yarn run estimate-on-premise-data --onPremiseInput <Input File Name>`. You can optionally add the argument `--onPremiseOutput <Output File Name>`.
+
+## Testing
+
+Using the [example .csv file](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/cli/src/__tests__/EstimateOnPremiseData/on_premise_data_input.test.csv) provided by CCF, the command line tool successfully created a corresponding estimations csv. This covers a period of 7 days, with 4 hours of daily uptime and 12 hours of weekly uptime, which suggests a system running 3 days a week. The monthly and annual uptime are both set to 36, which seems a little confusing. This possibly represents an example where only a monthly estimation was desired, but the tool forces all uptime types to be populated. The values are multiples of the daily calculated amounts so there does not appear to be any additional benefit to this other than convenience.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: You've only used ".csv" elsewhere, so I assume you meant to here.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+As I started to test out the tool using some generated .csv files, I often found that the process would fail. Very minimal information was given about what went wrong, the error message was simply: `Something went wrong: Input data is incorrect. Please check your input data file and try again.`. This wasn’t very helpful and necessitated investigating the source code to resolve the problem, which I will cover later.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: "...what went wrong; the error...", or "...what went wrong, as the error..."

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+The .csv input format does not appear to be an exact match for any standard input source and so required some additional work to prepare data for estimation. Thankfully this wasn’t too extensive a task, and with the help of some Python scripting using the [Pandas](https://pandas.pydata.org/) library, I was able to transform an exported SCCM spreadsheet into the minimum required data.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Since you're immediately following this with the script, I'd suggest ending this line with a colon.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+### Data Validation
+
+The input data is first validated, checking that each row has the required columns and throwing an error if any are missing or contain an empty string. This turned out to be the problem with my generated .csv files, where not all entries were populated. Adding the row detail to the error at this point would be extremely helpful in identifying why a particular input file causes problems, especially when dealing with a large amount of data.
+
+There was less extensive validation than I might have expected, like ensuring that the dates are in the required format. This may relate to the dates being formatted incorrectly in the output file.
+
+### Report Generation
+
+Apart from an entry function on the main App, the reporting is isolated in an on-premise package. It has a unique OnPremiseDataReport type, which is passed a standard ComputeEstimator and MemoryEstimator. These are the same classes that are used in the cloud calculations but here a specific on-premise memory coefficient is given to the MemoryEstimator. Currently the on-premise and all cloud providers use a value of 0.000392kWh / Gb but the system is setup so that they could be different.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: "set up", since it's a verb phrase rather than a noun phrase.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+#### Intel
+
+The Intel descriptions have a more complex process which attempts to remove ‘Intel(R)‘, ‘CPU’, and its clock speed from the description to leave a processor code only. Again, from my sample data, a description like ‘Intel(R) Core(TM) i5-7500 CPU @ 3.40GHz‘ should result in a lookup string of ‘i5-7500’ (which is also not supported). Others leave a ‘v2’ on the end preventing a match and a few like ‘Intel Pentium III Xeon processor’ do not produce a sensible key at all.
+
+When neither manufacturer produces a result, an on-premise specific average is used instead. Also, even though a common processor type can be determined, these again have custom coefficients for on-premises usage.
+
+### Region Data
+
+The input requires a country and a region but unless the country is 'United States' then the region is discarded. If it is, then the two are combined into a single string like 'United States-Texas'. This is then used in another on-premise specific lookup table to find the regional carbon intensity. There is an 'Unknown' region, which contains an average value that the code falls back on if the given region cannot be found.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Something like "If the region is used, then..." in place of "If it is"; it currently reads like region is used if it's discarded.
Below, "Unknown" is given as a country, not a region.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+| Romania | 0.00026184 |
+| South Korea | 0.0004156 |
+| Spain | 0.00017103 |
+| Sweden | 0.00000567 |
+| Switzerland | 0.00001152 |
+| United Kingdom | 0.00021233 |
+| United States | 0.00042394 |
+| Unknown | 0.0003228315385 |
+
+With the following regions supported for United States:
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: As a separate sentence, I'd suggest "The following regions are supported..." rather than "With the...".

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+| Texas | 0.00041432 |
+| Washington | 0.00013567 |
+| Ohio | 0.00056357 |
+| Oregon | 0.00017562 |
+
+Given that the validation forces you to put something in the region column, there is no way to use the base United States value on its own. If the region is not found in the above list, then it will again fall back on the global 'Unknown' average.
+
+### Machine Type
+
+The machine type can be specified as either server, desktop, or laptop. This can influence the CPU utilization and average watts used in calculations. If the type does not match one of these then this results in the CPU utilization being passed through (either specified on the row or the 50% default) and the Average watts being undefined, so that it is later calculated via:
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: "...these then the CPU utilization is passed..." would be simpler.
Did you mean to capitalise "Average" here?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+
+Average Watts = Min Watts + (average CPU utilization / 100) \* (Max Watts - Min Watts)
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I'd suggest either capitalising everything (well, not articles or pronouns), or nothing, regardless of your stance in body text.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        AVERAGE_WATTS?: number
+    }
+    DESKTOP?: {
+        CPU_UTILIZATION?: number
+        AVERAGE_WATTS?: number
+    }
+}
+~~~
+
+Then the CPU utilization and average watts will be overridden by these config values, skipping the calculation in the Compute Estimator. This seems a little surprising as it would be more natural to use these as fallback defaults, which would only be used if the CPU utilization was not specified on any given row.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I'd prefer "In this case, the CPU utilization...".

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+
+### Start and End Times
+
+These are required fields but do not appear to have any meaningful use currently. As mentioned in the validation section, the format is not checked and appears to only retain the hours, minutes, and milliseconds in the output data.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: "... not checked and the fields appear to only ..." would be better.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+## Integration with the web application
+
+At present the CLI tool is the only consumer of the on-premise reporting. While some of the Estimator classes are re-used, there is a lot of custom setup to use them, and on-premise specific coefficients used for computation. There is a [closed issue](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/issues/833) mentioning this but I don’t see any other indication that this is currently planned on the project backlog.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: "... to use them, or the on-premise ..."; if you have to use both together, delete this comma instead.
"... indication that any changes are planned..." or similar; the object is unstated otherwise.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+- Allow for more flexibility in the input country (use country codes etc.).
+- Improve the code that breaks down CPU descriptions so that more of my sample data would match the given types (a Regex would probably be more consistent).
+- Allow for sub regions outside of the United States (<https://carbonintensity.org.uk/> supports UK Postcodes and other named regions).
+- Change the configured machine type defaults to be used as a fallback only when CPU utilization is not specified.
+
+### Move the on-premise code to be more integrated with the cloud measurements.
+- Allow for usage of electricity maps and other live APIs.
+- Make use of start and end times to get specific Carbon Intensity if provided.
+- Unify the on-premise specific Carbon Intensity values with those used for cloud estimation. Should allow the list to expand and allow easier maintenance.
+- Same for identification of CPUS.
```
Comment (abalkham) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I'd use "CPUs".

**Review by abalkham — APPROVED** [review]
Nothing big wrong here; just a small pile of suggestions to improve grammar and readability.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+date: 2024-05-28 00:00:00 Z
+categories:
+- Sustainability
+author: mgriffin
+summary: The Cloud Carbon Footprint tool also provides support for carbon calculations from On-Premise sources. I investigate how it works and list some pros and cons with its usage, along with some contributions we are hoping to make.
+---
+
+## Introduction
+
+The Cloud Carbon Footprint tool was discussed in a [previous blog post](https://blog.scottlogic.com/2023/10/19/tools-for-measuring-cloud-carbon-emissions.html), as a third party method of estimating emissions associated with cloud workloads. However, it can also use data from [On-Premise](https://www.cloudcarbonfootprint.org/docs/on-premise) sources for its estimations. This blog post gives an overview of how it should be used and investigates how it works via testing and analysis of its code. I consider the issues with its use and list some contributions that could be made to improve it.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: This gives an accurate picture of what's in the blog post, but I think it would be even better with a quick summary of why we should care. A couple of ideas spring to mind:

* You want to manage computing-related carbon emissions to meet your enterprise's Net Zero target
* You keep your enterprise's data on-prem
* On-prem compute has significantly different carbon emissions compared to cloud, so you want to take this into account when you produce your carbon emissions estimates

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+## Usage Overview
+
+The system requires a .csv file as an input, which must contain a minimum set of columns, with Machine Name, cost, power usage effectiveness and CPU utilization being optional. If CPU utilization is not specified then it is assumed to be 50% by default; it can also be configured to use a different amount for server, desktop, and laptop machine types.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I personally found this a little difficult to follow -- granted, I haven't read the previous blog post, but I think even readers who know what to expect would find it easier to follow with some clear reminders of what this CSV file is and what it's for. I think using bullet points would also make it much easier to see at a scan what the columns of this CSV ought to be.

E.g.

```suggestion
The tool will estimate your carbon emissions for you, but in order to do this, it needs to know all about your compute usage. This information is provided as a CSV file with the following columns:

- [Compulsory column 1]
- [Compulsory column 2]
- [Compulsory column 3]
- Machine Name (optional)
- Cost (optional)
- Power Usage Effectiveness (optional)
- CPU utilization (optional, default: 50%)

You can also configure the tool to use a different CPU utilization value for different types of machine, such as servers, desktops and laptops.
```

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+This .csv file must then be passed to a command, which generates an output .csv file with estimated energy usage and CO2e emissions. This command is `yarn run estimate-on-premise-data --onPremiseInput <Input File Name>`. You can optionally add the argument `--onPremiseOutput <Output File Name>`.
+
+## Testing
+
+Using the [example .csv file](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/packages/cli/src/__tests__/EstimateOnPremiseData/on_premise_data_input.test.csv) provided by CCF, the command line tool successfully created a corresponding estimations .csv. This covers a period of 7 days, with 4 hours of daily uptime and 12 hours of weekly uptime, which suggests a system running 3 days a week. The monthly and annual uptime are both set to 36, which seems a little confusing. This possibly represents an example where only a monthly estimation was desired, but the tool forces all uptime types to be populated. The values are multiples of the daily calculated amounts so there does not appear to be any additional benefit to this other than convenience.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: What is this paragraph about? Is it just a list of strange behaviours you discovered, which readers might want to watch out for? If so, a good topic sentence to explain this would help.

Have you considered breaking this paragraph up into multiple shorter paragraphs, or even a bullet-point list?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+## Usage Overview
+
+The system requires a .csv file as an input, which must contain a minimum set of columns, with Machine Name, cost, power usage effectiveness and CPU utilization being optional. If CPU utilization is not specified then it is assumed to be 50% by default; it can also be configured to use a different amount for server, desktop, and laptop machine types.
+
+Each row has a start and end time that covers the period it relates to, as well as daily, weekly, monthly, and annual uptime amounts in hours. This allows a row to cover a period where the hardware is only active some of the time. For example, you could generate a weekly .csv file for your company laptop which has eight daily hours, forty weekly hours etc.
+
+This .csv file must then be passed to a command, which generates an output .csv file with estimated energy usage and CO2e emissions. This command is `yarn run estimate-on-premise-data --onPremiseInput <Input File Name>`. You can optionally add the argument `--onPremiseOutput <Output File Name>`.
+
+## Testing
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I feel this section title could be better. Is this a guide to testing the tool? Or a guide to testing your applications for carbon efficiency using the tool? I'd prefer a title which tells me roughly what's underneath it, and which gives me a reason why I should read it.

Perhaps:

```suggestion
## Gotchas to look out for
```

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+
+if __name__ == "__main__":
+    input_excel_file = "input.xlsx"
+    output_csv_file = "output.csv"
+    sheet_name = "sheet name"
+
+    main(input_excel_file, output_csv_file, sheet_name)
+~~~
+
+## Code Analysis
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Are you sure you want this to be a separate section from the previous one? I get the sense that they are both about 'gotchas' with the tool. If so, I feel a title like `## Gotchas` should cover them both

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+
+### Input/Output
+
+Despite the documentation stating that the input must be placed in the `packages/cli` folder of the tool, it uses the path directly so you can pass in an absolute path from outside of the tool. However, the output file name is combined with the process' current working directory, so there is not currently a way to control the output location.
+
+It also uses the same default output file name each time so care may need to be taken to ensure that results are not overwritten. Because of this, another error that can occur is having the previous .csv file open in Excel, which prevents the new file from being written.
+
+The whole csv file is read into memory at once, using the 'csvtojson' library – this was not a problem in my test cases but is something to bear in mind if dealing with an extremely large amount of data.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: These are great. I think you can help readers out by putting the bit they're most likely to care about in the topic sentence. E.g.

```
There is no way to control the output directory. This is because...

You can easily overwrite your outputs if you run the tool several times in a row, because...

If you're using an extremely large amount of data, be aware that this may be very slow, because...
```

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+
+In this case, the CPU utilization and average watts will be overridden by these config values, skipping the calculation in the Compute Estimator. This seems a little surprising as it would be more natural to use these as fallback defaults, which would only be used if the CPU utilization was not specified on any given row.
+
+### Start and End Times
+
+These are required fields but do not appear to have any meaningful use currently. As mentioned in the validation section, the format is not checked and the fields appear to only retain the hours, minutes, and milliseconds in the output data.
+
+## Integration with the web application
+
+At present the CLI tool is the only consumer of the on-premise reporting. While some of the Estimator classes are re-used, there is a lot of custom setup to use them and on-premise specific coefficients used for computation. There is a [closed issue](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/issues/833) mentioning this but I don’t see any other indication that any changes are planned on the project backlog.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I'm a bit lost. You might want to consider re-writing this paragraph.

---

```
At present the CLI tool is the only consumer of the on-premise reporting.
```

I found this hard to follow because it uses abstract nouns. Could be easier to read if you pack the meaning into more concrete terms and into verbs instead. E.g.:

```
At present, you can only access the on-premise reports through the CLI, not the web app.
```

---

```
While some of the Estimator classes are re-used, there is a lot of custom setup
to use them and on-premise specific coefficients used for computation.
```

- Where are the Estimator classes `re-used`? In the code for both the CLI and for the web app interface?
- I'm confused how there can be `a lot of custom setup to use them`, since I'm assuming `them` to refer to the Estimator classes, and `custom setup` to mean the end user configuring their settings. But 'using classes' is surely something _developers_ do, not end users. Hence why I'm confused
- I can't see how this sentence is supposed to work grammatically, so I can't quite figure out what you meant to say

---

```
I don't see any other indication that changes are planned
```

Should we have expected that the repo maintainers would be planning changes? Should we want changes? What's the problem with the way things are now?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+
+## Usage Considerations
+
+### Positives
+
+- Easy to generate information in required format.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: `required` for what?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+- Does not block execution if given data does not match expectations.
+- Uses the same core estimation code.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: `the same` as what?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+
+### Downsides
+
+- Error messages are not obvious.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]:
```suggestion
- Error messages are too vague.
```

`not obvious` to me suggests that it's hard to find the error messages, as if the error log was well-hidden or something. I'm guessing this isn't what you meant?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+- Little transparency on when fallback averages have been used.
+- Lots of lists of values specific to the on-premise calculations.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: What's the problem here?

- You, the end user, have to specify lots of values manually, making it a huge faff to use?
- There are lots of hard-coded values, making the tool inflexible for diverse use cases?
- Something else?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+
+### Integrate the On-Premise code with the API/client.
+- Add an API endpoint to submit a CSV file that can be added to the data cache.
+  - To use pre-existing properties this could potentially list 'On-Premise' as Cloud Provider, &lt;Machine Name&gt; as Account, &lt;Machine Type&gt; as Service etc.
+- Add client functionality to upload a CSV file.
+- Add client functionality to view on-premise data.
+
+## Conclusion
+
+This tool was helpful in getting a high-level estimate of the associated carbon emissions based on a large list of hardware resources. The fact that the tool seamlessly falls back on average values is beneficial in being able to use the tool easily. However, it could be seen as lacking in transparency around the values used for its calculations. Combined with the amount of hardware and countries that are not supported means I would not recommend it for accurate reporting in its current form.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: - I personally find the word `high-level` in this context hard to understand. Do you mean 'overall', as opposed to getting an estimate broken down per module or per region or whatever? Or do you mean 'approximate', as opposed to a precise estimate? Or both?
- You start this paragraph with the phrase, `This tool was helpful`. That sets up the reader with the expectation that you are _recommending_ this tool. But at the end of the paragraph, you do the opposite, and recommend _against_ the tool. You should make it clear where you're going with this paragraph in the topic sentence.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+
+The lack of integration with the web application means that you would need to turn to another process to visualise the output. Adding support for on-premise data in the web application should be possible but could be a more significant amount of work than some other identified improvements.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]:
```suggestion
The tool doesn't integrate with the web application. This means that you would need to turn to another process to visualise the output. Adding support for on-premise data in the web application should be possible but could be a more significant amount of work than some other identified improvements.
```

Again, just another wee opportunity to pack meaning into concrete verbs rather than abstract nouns.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+
+I should caveat that I understand that a tool designed for measuring your 'cloud' carbon footprint might not have mature capabilities for on-premise measurement. This is an area without clear established standard tooling but is still an important aspect of many companies carbon footprint.
```
Comment (jcarstairs-scottlogic) [inline _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]:
```suggestion
I should caveat that I understand that a tool designed for measuring your 'cloud' carbon footprint might not have mature capabilities for on-premise measurement. This is an area without clear established standard tooling but is still an important aspect of many companies’ carbon footprint.
```

**Review by jcarstairs-scottlogic — COMMENTED** [review]
Nice work Matthew, I'm sure this will be very useful for people interested in this tool.

I've just got a few suggestions for bits I found confusing as a non-expert. It may well be that you're writing for a more informed audience than me, so feel free to politely ignore all my comments!

One general comment is that you should be clear from the first sentence what this blog post is about -- which is to say, why I should bother reading it?

- Is it to help the reader use this tool for themselves?
- Is it to help them evaluate whether to use this tool over an alternative?
- Is it something else?

Possibly this post does more than one of these things, which will irritate my tidy-mindedness, but it's still better to communicate that right from the start. You could even try out some different title ideas to see if they get the message across better.

**Comment by mgriffin-scottlogic** [comment]
> Nice work Matthew, I'm sure this will be very useful for people interested in this tool.
> 
> I've just got a few suggestions for bits I found confusing as a non-expert. It may well be that you're writing for a more informed audience than me, so feel free to politely ignore all my comments!
> 
> One general comment is that you should be clear from the first sentence what this blog post is about -- which is to say, why I should bother reading it?
> 
> * Is it to help the reader use this tool for themselves?
> * Is it to help them evaluate whether to use this tool over an alternative?
> * Is it something else?
> 
> Possibly this post does more than one of these things, which will irritate my tidy-mindedness, but it's still better to communicate that right from the start. You could even try out some different title ideas to see if they get the message across better.

Thanks a lot Joe, lots of good stuff to consider here. In terms of what it's about, I guess it grew out of an internal document to examine the tool and whether we would use it, so maybe the first point mostly? At the time it was written I wouldn't say there was a clear alternative, certainly not that was free and open source anyway. But I do think it's a very different prospect from their automated cloud based solution, which might not be what everyone would hope for.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 14 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Thanks, I've kind of included all of these, if that's not too much?

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 15 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Added this, my only concern was it was a long list, which is displayed as a table on their doc page.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 16 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Tried to incorporate some of the previous paragraph to make this into a 'Getting Started' section and move the strangeness into 'Gotchas'

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 20 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: Hope this makes more sense now, I just dropped the section about custom setup. It was more from my position as a coder not liking the amount of duplication 😄 
Also talked about how you might want to visualise and talk about it being on their backlog at one point.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 24 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I guess this is a little bit of something else, again looking at it as a coder and thinking that there are unnecessary duplicate values. I could be hard to maintain two lists of essentially the same information.

File: _posts/2024-05-28-cloud-carbon-footprint-on-premise.md (hunk 25 above)
Comment (mgriffin-scottlogic) [reply _posts/2024-05-28-cloud-carbon-footprint-on-premise.md]: I think I mean approximate and have put the reservations first (as well as split the paragraph)

**Review by jcarstairs-scottlogic — APPROVED** [review]
